### PR TITLE
kill: add a feature decoding signal masks

### PR DESCRIPTION
**Pull request by masatake**
-d/--decode-mask option works well with the output of ps command:
```
   $ ps s $$
     UID     PID          PENDING          BLOCKED          IGNORED           CAUGHT STAT TTY        TIME COMMAND
    1000 1588189 0000000000000000 0000000000010000 0000000000384004 000000004b813efb S    pts/56     0:00 bash

   $ ./kill --decode-mask 000000004b813efb
   HUP
   INT
   ILL
   TRAP
   ABRT
   ...
```
If you know the pid you are interested in, you can skip running ps comman with -D/--decode-process-state option:
```
   $ ./kill --decode-process-state $$
   SigPnd:
   ShdPnd:
   SigBlk:
	   CHLD
   SigIgn:
	   QUIT
	   TERM
	   TSTP
	   TTIN
	   TTOU
   SigCgt:
	   HUP
	   INT
	   ILL
	   TRAP
	   ABRT
	   BUS
	   FPE
	   USR1
           ...
```

- [x] add test cases

### CODE REVIEW

**Comment by karelzak** [comment]
Ah, I like this idea :-) 

What do you think about adding support for decoding the masks to the --list? It already supports conversion from signal number to signal name (`--list 9`) and also supports masks prefixed by 0x seems like a native thing; `kill --list 0x000000004b813efb`.

And `--show-process-state` rather than `--decode-process-state`, and report it in more human-readable way
* non-empty items only
* use human-readable names for the masks (SigBlk --> Blocking)
* not separate signals by \n
```
$ kill --show-process-state $$`
Blocking: CHLD
Ignoring: QUIT TERM TSTP TTIN TTOU
Catching: HUP INT ILL TRAP ABRT BUS FPE USR1
```

I can also imagine `lssig` listing all about signals from /proc :-)))

**Comment by masatake** [comment]
@karelzak Thank you for the comment. I updated the change:

```
       $ ps s $$
         UID     PID          PENDING          BLOCKED          IGNORED           CAUGHT STAT TTY        TIME COMMAND
        1000 1588189 0000000000000000 0000000000010000 0000000000384004 000000004b813efb S    pts/56     0:00 bash
    
       $ ./kill --list 0x000000004b813efb
       HUP
       INT
       ILL
       TRAP
       ABRT
       ...
       $ ./kill --show-process-state 1
       Blocked: HUP INT USR1 USR2 TERM CHLD WINCH PWR RT0 RT1 RT2 RT3 RT4 RT5 RT6 RT7 RT13 RT14 RT15 RT16 RT17 RT18 RT20 RT21 RT22 RT23 RT24 RT25 RT26 RT27 RT28 RT29 
       Ignored: PIPE 
       Caught: QUIT ILL ABRT BUS FPE SEGV       
```

> I can also imagine lssig listing all about signals from /proc :-)))

Eventually, we will want to implement `lsps` with `-o` and `-Q` options.

File: misc-utils/kill.c [hunk 1]
```diff
@@ -286,24 +354,58 @@ static char **parse_arguments(int argc, char **argv, struct kill_control *ctl)
...
 		if (!strcmp(arg, "-L") || !strcmp(arg, "--table")) {
 			print_all_signals(stdout, 1);
 			exit(EXIT_SUCCESS);
 		}
+		if (!strcmp(arg, "-d") || !strcmp(arg, "--show-process-state")) {
+			uint64_t pid;
+			if (argc < 2)
+				errx(EXIT_FAILURE, _("too few arguments"));
+			if (2 < argc)
+				errx(EXIT_FAILURE, _("too many arguments"));
```
Comment (t-8ch) [inline misc-utils/kill.c:394]: `if (argc != 2)`

File: misc-utils/kill.c [hunk 2]
```diff
@@ -286,24 +354,58 @@ static char **parse_arguments(int argc, char **argv, struct kill_control *ctl)
...
-			print_signal_name(ctl->numsig);
+			print_signal_name(ctl->numsig, true);
 			exit(EXIT_SUCCESS);
 		}
 		if (!strcmp(arg, "-L") || !strcmp(arg, "--table")) {
 			print_all_signals(stdout, 1);
 			exit(EXIT_SUCCESS);
 		}
+		if (!strcmp(arg, "-d") || !strcmp(arg, "--show-process-state")) {
+			uint64_t pid;
```
Comment (t-8ch) [inline misc-utils/kill.c:390]: Why not `pid_t`?

File: misc-utils/kill.c [hunk 3]
```diff
@@ -103,21 +104,84 @@ struct kill_control {
...
+	}
+}
+
+static void print_process_signal_state(pid_t pid)
+{
+	struct path_cxt *pc = NULL;
+	FILE *fp;
+	char buf[BUFSIZ];
+
+	const struct sigfield {
```
Comment (t-8ch) [inline misc-utils/kill.c:142]: `static`

File: misc-utils/kill.c [hunk 4]
```diff
@@ -103,21 +104,84 @@ struct kill_control {
...
+		err(EXIT_FAILURE, _("cannot open /proc/%d/status"), pid);
+
+	while (fgets(buf, sizeof(buf), fp) != NULL) {
+		for (size_t i = 0; i < ARRAY_SIZE(sigfields); i++) {
+			const char *key = sigfields[i].key;
+			size_t keylen = strlen(key);
+
+			if (strncmp(buf, key, keylen) == 0) {
+				char *val = buf + keylen;
+				uint64_t sigmask;
```
Comment (t-8ch) [inline misc-utils/kill.c:167]: Aren't sigmasks `int`?

Comment (masatake) [reply misc-utils/kill.c:167]: At least on x86_64, int looks too small.
```
[yamato@dev64]~/var/util-linux/util-linux% cat /tmp/foo.c
int main(void) {return sizeof(int);}
[yamato@dev64]~/var/util-linux/util-linux% gcc /tmp/foo.c
[yamato@dev64]~/var/util-linux/util-linux% ./a.out; echo $?
4
[yamato@dev64]~/var/util-linux/util-linux% ./kill -l 0xffffffffffffffff | cat -n | tail
    55	RT21
    56	RT22
    57	RT23
    58	RT24
    59	RT25
    60	RT26
    61	RT27
    62	RT28
    63	RT29
    64	RT30
```

Am I missing something?

File: misc-utils/kill.c (hunk 3 above)
Comment (masatake) [reply misc-utils/kill.c:142]: Thank you. Fixed in locally.

File: misc-utils/kill.c (hunk 2 above)
Comment (masatake) [reply misc-utils/kill.c:390]: Thank you. Fixed in locally.

File: misc-utils/kill.c (hunk 4 above)
Comment (t-8ch) [reply misc-utils/kill.c:167]: Ah, I looked at `sigmask(3)` which is deprecated.

pthread_sigmask(3) proposes `sigset_t`. But those are 128 *bytes* big on amd64 glibc and a structure to boot.
So a `uint64_t` is probably the best to work with.

File: misc-utils/kill.c [hunk 5]
```diff
@@ -103,21 +104,84 @@ struct kill_control {
...
+
+	const struct sigfield {
+		char *key;
+		char *label;
+	} sigfields[] = {
+		{ "SigPnd:\t", "Pending (thread)"  },
+		{ "ShdPnd:\t", "Pending (process)" },
+		{ "SigBlk:\t", "Blocked"           },
+		{ "SigIgn:\t", "Ignored"           },
+		{ "SigCgt:\t", "Caught"            },
```
Comment (karelzak) [inline misc-utils/kill.c:150]: Let's translate the strings:
```
   N_("Caught")
```

File: misc-utils/kill.c [hunk 6]
```diff
@@ -103,21 +104,84 @@ struct kill_control {
...
+				char *val = buf + keylen;
+				uint64_t sigmask;
+
+				rtrim_whitespace((unsigned char*)val);
+				if (ul_strtou64(val, &sigmask, 16) < 0) {
+					warnx( _("unexpected sigmask format: %s (%s)"), val, key);
+					continue;
+				}
+				if (sigmask != 0) {
+					printf("%s: ", sigfields[i].label);
```
Comment (karelzak) [inline misc-utils/kill.c:175]: printf("%s: ", _(sigfields[i].label));

File: misc-utils/kill.c [hunk 7]
```diff
@@ -211,9 +275,13 @@ static void __attribute__((__noreturn__)) usage(void)
 		"                        wait up to timeout and send follow-up signal\n"), out);
 #endif
 	fputs(_(" -p, --pid              print pids without signaling them\n"), out);
-	fputs(_(" -l, --list[=<signal>]  list signal names, or convert a signal number to a name\n"), out);
+	fputs(_(" -l, --list[=<signal>|=0x<sigmask>]\n"
+		"                        list signal names, convert a signal number to a name,\n"
+		"                         or convrt a signal mask to names\n"), out);
 	fputs(_(" -L, --table            list signal names and numbers\n"), out);
 	fputs(_(" -r, --require-handler  do not send signal if signal handler is not present\n"), out);
+	fputs(_(" -d, --decode-process-state <pid>\n"
```
Comment (karelzak) [inline misc-utils/kill.c:283]: --show-process-state

File: misc-utils/kill.c [hunk 8]
```diff
@@ -286,24 +354,58 @@ static char **parse_arguments(int argc, char **argv, struct kill_control *ctl)
...
 		}
+		if (!strcmp(arg, "-d") || !strcmp(arg, "--show-process-state")) {
+			uint64_t pid;
+			if (argc < 2)
+				errx(EXIT_FAILURE, _("too few arguments"));
+			if (2 < argc)
+				errx(EXIT_FAILURE, _("too many arguments"));
+			arg = argv[1];
+			if (ul_strtou64(arg, &pid, 10) < 0)
+				errx(EXIT_FAILURE, _("wrong pid specification: %s"), arg);
```
Comment (karelzak) [inline misc-utils/kill.c:397]: pid = strtou64_or_err(argv[1], _("invalid pid argument"));

Comment (t-8ch) [reply misc-utils/kill.c:397]: With the change to `pid_t`: `strtopid_or_err()`

Comment (masatake) [reply misc-utils/kill.c:397]: I didn't know strtopid_or_err. Thank you.

File: misc-utils/kill.c (hunk 5 above)
Comment (masatake) [reply misc-utils/kill.c:150]: Fixed in locally.

File: misc-utils/kill.c (hunk 6 above)
Comment (masatake) [reply misc-utils/kill.c:175]: Fixed in locally.

File: misc-utils/kill.c (hunk 7 above)
Comment (masatake) [reply misc-utils/kill.c:283]: Fixed in locally.

File: misc-utils/kill.c (hunk 8 above)
Comment (masatake) [reply misc-utils/kill.c:397]: Fixed in locally.

**Comment by masatake** [comment]
lsns/filter test case failed on rpm-build:fedora-rawhide-ppc64le. Mysterious. I will record the failure in #2967.

**Comment by masatake** [comment]
Added a test case.

**Pull request merged**